### ISSUE TITLE
Questions about the paper

### ISSUE THREAD

**Issue by realquantumcookie** [SakanaAI/continuous-thought-machines#9]
Dear authors:

Thank you for such thoughtful paper, drawing inspirations from the neural science field. I'm very intrigued by your idea, and want to ask some follow-up questions after reading your paper:

1. The loss used by your paper tries to optimize the average of (1) the loss at the tick when the model preforms the best (2) the loss at the tick when the model is most certain. I believe an important aspect for your paper is the flexibility to scale up / down compute in exchange for better accuracy. But in the loss term I don't seem to see anything that encourages model outputs to converge in earlier ticks, so is this a natural behavior that emerged during training?
2. During training I assumed you fixed the number of ticks used to train a model?
3. Is the synchronization model used for allowing arbitrary routing of neuron outputs / inputs? I'm not quite familiar with this setting and would like to learn any motivation for it. Also if we remove the synchronization model and just modify MLPs to include pre-activation histories, what will be the result (I assume you guys also tried this internally?)

Thank you for spending your time answering my questions, I do think the emerging behaviors in your visualizations look cool and this thing looks very natural to the human brain.

**Comment by lukedarlow** [comment]
Hello, and thank you for your questions. 

**I would appreciate if you opened this as a github 'discussion' instead of an issue if you wish to continue engaging on this, as I think this is more of a discussion than a real 'issue', so to speak.**

Before I get to answering your questions, I want to put forward that all of the components of the CTM seem to work in synergy. We spent months of experimentation to unpack and understand the functionality of various components, discarding dozens of approaches and ideas before settling on the current structure. At first glance, the novel components (internal ticks to think with, NLMs, and synchronization) seem unusual, but they all work together in a unique way. I mention this because, for instance, answering your question about the loss function is somewhat contingent in understanding the deeper role of synchronization. I shall do my best! 

## Re: loss function
You correctly assume that there is no 'driving' mechanism to minimise the number of ticks, so I shall share my intuition and expectations to better elucidate an understanding. 

First, yes, the adaptive compute time component falls out as a (semi) emergent property. I use the term 'semi' as a precursor there because it depends on a number of factors and/or design choices. For instance, if you have far fewer classes than internal ticks, the behavior of the predictions can be very different. By way of example, consider CIFAR-10 (with 10 classes). If you use 10 (or more) internal ticks the CTM (or any model for that matter) could learn to output each class at each internal tick, meaning that what the model needs to do to minimise the loss is put probability mass into the class at the exact tick that it thinks is correct (this is sort of a 'shortcut' approach that is typical to NNs). But, if you have many more classes or a more challenging task (e.g., ImageNet or the maze task), you end up with exactly what you would expect from adaptive compute emerging. 

Second, synchronization is well-suited to this. The way that we compute synchronization is as a dot-product between post-activation neuron pair histories. Therefore, as those histories grow longer, the **absolute size** of the dot product can grow larger too (although it is normalized by a sqrt factor, it can still grow). Intuitively this is a mechanism for 'becoming more sure' over time - the more time you are synchronizing, the more synchronized you can become. It is, in some sense, an attractor-like state, where a natural side-effect is increasing certainty. 

Third, my intuition for the **inductive bias** regarding internal ticks and adaptive compute in the context of the CTM is as follows. The CTM can 'spread' the data it classifies over the internal tick dimension (I view it as a 'thought dimension') as this is the best way to optimise the underlying task objective. The 'best spread' is that which attributes more compute to harder examples. So, from an inductive bias perspective, one would expect what we are beginning to call **'Native Adaptive Compute'** as a by-product.

## Re: fixed number of internal ticks?
Yes. 

This is a fascinating area and something that you should certainly explore!

## Re: Synchronization routing

I don't quite understand your question about arbitrary routing. Could you open a discussion (port your original question and my answer, if you like!) and go into more depth on that?

## Re: discarding synchronization?

**Synchronization does something rather special: it mitigates the need for time alignment**. What I mean by this is that synchronization (merely dot products between histories) yields a representation where the underlying units that make this representation (i.e., neural _histories_) are not required to be strictly time-aligned. They can each have the freedom to be far more dynamic and open-ended, which ties in nicely to the biological motivation behind this work. They can all exhibit varied and compound periodicities, or none. The NLMs can learn to do something simple or complex, or anything in between! In the paper we mentioned how using internal ticks and NLMs the way we do actually introduced this issue of time alignment. It was very gratifying that another biologically plausible idea was an idea solution.

What we tried and why it didn't work:

1. Computing q+preds directly off of **z**. It doesn't work as well because it constrains the variability of the neurons themselves. Consider that if a neural was truly 'dynamic' it would be changing value often (going up and down, just like the gif at the top of the github repo readme). This change would dramatically change the structure of the **z** representation over time. Projecting directly off of that would either very much limit the dynamics (it does, we tested this) or it would create an obvious periodic behavior in the q+preds (it does, we tested this). Neither of those outcomes is great to optimise with. 
2. Building a 'holder representation'. Instead of directly projecting q+preds off of **z**, keep an intermediate 'holder' representation that you project into and accumulate information, perhaps with a more powerful projection. Then project off of that to q+preds. This alleviates the problems in 1, but they are still present. 
3. Use a smarter projection strategy such as a 1D convolution. Again, this alleviates 1 and 2 even more, but it is (1) parameter hungry, (2) has many difficult-to-tune hyperparams, and (3) not easy to motivate as biologically plausible (it looks very 'hacky'!)


Your suggestion of just using MLPs off of pre-activation histories is very similar to 3, but introduces additional problems of time-alignment. What I mean by this is projecting to q+preds as you suggested would work fine if you only projected at a pre-defined single internal tick as the underlying models that do the projection would be biased to learning specific structure for the incoming time series. If, however, you project off of every single tick as we do, you are requiring this underlying projection model to do a LOT more work as it should handle time series of various alignments with the model weights.

------------
I hope that helps your understanding. Let me know if I can assist further.
